Add support for an optional login and document level access control system. (#624)
Azure-Samples/azure-search-openai-demo: LoginAndAclSetup.md

# Setting up optional login and document level access control

## Table of Contents

- [Requirements](#requirements)
- [Setting up Azure AD Apps](#setting-up-azure-ad-apps)

>   - [Manual Setup](#manual-setup)

Author: Unfortunately missing from this PR - automatic setup. It's quite the can of worms because it involves calling the graph API, but we are now pulling in the graph sdk which should make this simpler. Can automatic setup wait until a follow up PR? @pamelafox  has a similar script in her PR #534

Author: follow up pr

This guide demonstrates how to add an optional login and document level access control system to the sample. This system can be used to restrict access to indexed data to specific users based on what [Azure Active Directory (Azure AD) groups](https://learn.microsoft.com/azure/active-directory/fundamentals/how-to-manage-groups) they are a part of, or their [user object id](https://learn.microsoft.com/partner-center/find-ids-and-domain-names#find-the-user-object-id).

![AppLoginArchitecture](./docs/applogincomponents.png)

## Requirements

**IMPORTANT:** In order to add optional login and document level access control, you'll need the following in addition to the normal sample requirements

* **Azure account permissions**: Your Azure account must have [permission to manage applications in Azure AD](https://learn.microsoft.com/azure/active-directory/develop/quickstart-register-app#prerequisites).

## Setting up Azure AD Apps

Two Azure AD apps must be registered in order to make the optional login and document level access control system work correctly.

### Manual Setup

The following instructions explain how to setup the two apps using the Azure Portal.

> #### Setting up the Server App

Reviewer: I think my AAD Auth PR automates a lot of this. I prefer automation when possible as its a pain to setup app registrations and even more of a pain to debug the process (when an env variable isn't quite right). Let's join forces!

Author: We should do automation, agreed. The problem is setting up the API permissions on the apps, we need to fill in some of the body to the API calls with a bunch of manifest settings. and then we need to go back after the app is provisioned and setup the redirect uris correctly

Reviewer: I don't think my branch does the manifest setting, but it does include a postprovision hook to setup the redirect URIs. I can look into the manifest setting automation.

Author: we spoke offline - let's get this in a follow up pr

  1. Run the following `azd` command to save this ID: `azd env set AZURE_SERVER_APP_ID <Application (client) ID>`.
1. Select **Certificates & secrets** in the left hand menu.
1. In the **Client secrets** section, select **New client secret**.
  1. Type a description, for example `Azure Search OpenAI Demo Key`.
  1. Select one of the available key durations.
  1. The generated key value will be displayed after you select **Add**.
  1. Copy the generated ke value and run the following `azd` command to save this ID: `azd env set AZURE_SERVER_APP_SECRET <generated key value>`.
1. Select **API Permissions** in the left hand menu. By default, the [delegated `User.Read`](https://learn.microsoft.com/graph/permissions-reference#user-permissions) permission should be present. This permission is required to read the signed-in user's profile to get the security information used for document level access control. If this permission is not present, it needs to be added to the application.
  1. Select **Add a permission**, and then **Microsoft Graph**.
  1. Select **Delegated permissions**.
  1. Search for and and select `User.Read`.
  1. Select **Add permissions**.

> 1. Select **Expose an API** in the left hand menu. The server app works by using the [https://learn.microsoft.com/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow#protocol-diagram](On-Behalf-Of Flow), which requires the server app to expose at least 1 API.

Author: this link does not render properly. Fix using github markdown syntax

    1. Leave **State** set to **Enabled**.
    1. Select **Add scope** at the bottom to save the scope.

#### Client App

1. Sign in to the [Azure portal](https://portal.azure.com/).
1. Select the Azure AD Service.
1. In the left hand menu, select **Application Registrations**.
1. Select **New Registration**.
  1. In the **Name** section, enter a meaningful application name. This name will be displayed to users of the app, for example `Azure Search OpenAI Demo Web App`.
  1. Under **Supported account types**, select **Accounts in this organizational directory only**.
  1. Under `Redirect URI (optional)` section, select `Single-page application (SPA)` in the combo-box and enter the following redirect URI:
    1. If you are running the sample locally, use `http://localhost:50505/redirect`.
    1. If you are running the sample, use the endpoint provided by `azd up`: `https://<your-endpoint>.azurewebsites.net/redirect`.
1. Select **Register** to create the application
1. In the app's registration screen, find the **Application (client) ID**.
  1. Run the following `azd` command to save this ID: `azd env set AZURE_CLIENT_APP_ID <Application (client) ID>`.

> 1. In the left hand menu, select **API permissions**. You will add permission to access the **access_as_user** API on the server app. This permission is required for the [https://learn.microsoft.com/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow#protocol-diagram](On-Behalf-Of Flow) to work.

Author: this link does not render properly. Fix using github markdown syntax

  - [Manual Setup](#manual-setup)
    - [Server App](#server-app)
    - [Client App](#client-app)
- [Optional Scripts](#optional-scripts)
  - [Azure Data Lake Storage Gen2 Setup](#azure-data-lake-storage-gen2-setup)
  - [Azure Data Lake Storage Gen2 Prep Docs](#azure-data-lake-storage-gen2-prep-docs)
  - [Manually managing Document Level Access Control](#manually-managing-document-level-access-control)
- [Environment Variables Reference](#environment-variables-reference)

This guide demonstrates how to add an optional login and document level access control system to the sample. This system can be used to restrict access to indexed data to specific users based on what [Azure Active Directory (Azure AD) groups](https://learn.microsoft.com/azure/active-directory/fundamentals/how-to-manage-groups) they are a part of, or their [user object id](https://learn.microsoft.com/partner-center/find-ids-and-domain-names#find-the-user-object-id).

![AppLoginArchitecture](./docs/applogincomponents.png)

## Requirements

**IMPORTANT:** In order to add optional login and document level access control, you'll need the following in addition to the normal sample requirements

> * **Azure account permissions**: Your Azure account must have [permission to manage applications in Azure AD](https://learn.microsoft.com/azure/active-directory/develop/quickstart-register-app#prerequisites).

Reviewer: Would it be better to directly link to https://learn.microsoft.com/en-us/azure/active-directory/roles/permissions-reference#cloud-application-administrator ? That's what we do for the other roles in the main README

## Setting up Azure AD Apps

> Two Azure AD apps must be registered in order to make the optional login and document level access control system work correctly.

Reviewer: Is there a short explanation of why 2 apps are needed? I only needed one for mine.

Author: I am planning on going into more detail in the blog post - but the reason there's only 1 app needed for yours is that you are putting the login in front of the web app.

There's 2 apps here because the client UI is a Single Page Application and supports the login. This needs its own app. The API needs its own app to call the graph API

Author: updated the readme

### Manual Setup

The following instructions explain how to setup the two apps using the Azure Portal.

#### Setting up the Server App

* Sign in to the [Azure portal](https://portal.azure.com/).
* Select the Azure AD Service.
* In the left hand menu, select **Application Registrations**.
* Select **New Registration**.
  * In the **Name** section, enter a meaningful application name. This name will be displayed to users of the app, for example `Azure Search OpenAI Demo API`.
  * Under **Supported account types**, select **Accounts in this organizational directory only**.
* Select **Register** to create the application
* In the app's registration screen, find the **Application (client) ID**.
  * Run the following `azd` command to save this ID: `azd env set AZURE_SERVER_APP_ID <Application (client) ID>`.

> * Select **Certificates & secrets** in the left hand menu.

Reviewer: It will only be possible in Bicep to use certificates, and the docs all discourage the use of secrets, so I wonder if we should encourage the (admittingly more difficult) route of using a certificate here. Otherwise it'll be hard for folks to upgrade.

https://learn.microsoft.com/en-us/graph/applications-how-to-add-certificate

"Azure Active Directory (Azure AD) supports three types of credentials to authenticate apps and service principals: passwords (app secrets), certificates, and federated identity credentials. If you can't use federated identity credentials for your app, we strongly recommend that you use certificates instead of secrets."

The easiest way to get started locally with certificates is self-signed certificates, though those are also discouraged for production.

Author: So this makes it more challenging - I'll see if there's an easy way to add key vault support here. Are you OK leaving in the password for now and creating a follow up issue?

Reviewer: Yeah I'm fine with that. I'm very curious how we can empower developers to use certificates, based on my difficulties setting them up.

Author: Discussed offline - there is an opportunity to enhance this using Azure Key Vault + self-signed certs once we have bicep support.

Author: Added a note stating that certificates are considered more secure and linked to the certificate creation docs

  * Select **Add** to save your changes

#### Client App

* Sign in to the [Azure portal](https://portal.azure.com/).
* Select the Azure AD Service.
* In the left hand menu, select **Application Registrations**.
* Select **New Registration**.
  * In the **Name** section, enter a meaningful application name. This name will be displayed to users of the app, for example `Azure Search OpenAI Demo Web App`.
  * Under **Supported account types**, select **Accounts in this organizational directory only**.
  * Under `Redirect URI (optional)` section, select `Single-page application (SPA)` in the combo-box and enter the following redirect URI:
    * If you are running the sample locally, use `http://localhost:50505/redirect`.

>     * If you are running the sample, use the endpoint provided by `azd up`: `https://<your-endpoint>.azurewebsites.net/redirect`.

Reviewer: ```suggestion
    * If you are running the sample on Azure, use the endpoint provided by `azd up`: `https://<your-endpoint>.azurewebsites.net/redirect`.
```

>     * If you are running the sample from codespaces, use the codespaces endpoint: `https://<your-codespace>-50505.app.github.dev/`

Reviewer: ```suggestion
    * If you are running the sample from GitHub Codespaces, use the Codespaces endpoint: `https://<your-codespace>-50505.app.github.dev/`
```

* Select **Register** to create the application
* In the app's registration screen, find the **Application (client) ID**.
  * Run the following `azd` command to save this ID: `azd env set AZURE_CLIENT_APP_ID <Application (client) ID>`.
* In the left hand menu, select **Authentication**.
  * Under **Implicit grant and hybrid flows**, select **ID Tokens (used for implicit and hybrid flows)**
  * Select **Save**
* In the left hand menu, select **API permissions**. You will add permission to access the **access_as_user** API on the server app. This permission is required for the [On Behalf Of Flow](https://learn.microsoft.com/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow#protocol-diagram) to work.
  * Select **Add a permission**, and then **My APIs**.
  * In the list of applications, select your server application **Azure Search OpenAI Demo API**
  * Ensure **Delegated permissions** is selected.
  * In the **Select permissions** section, select the **access_as_user** permission
  * Select **Add permissions**.

#### Configure Server App Known Client Applications

> Consent from the user must be obtained for use of the client and server app. The client app can prompt the user for consent through a dialog when they log in. The server app has no ability to show a dialog for consent. Client apps can be [added to the list of known clients](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow#gaining-consent-for-the-middle-tier-application) to access the server app, so a consent dialog is shown for the server app.

Reviewer: ```suggestion
Consent from the user must be obtained for use of the client and server app. The client app can prompt the user for consent through a dialog when they log in. The server app has no ability to show a dialog for consent. Client apps can be [added to the list of known clients](https://learn.microsoft.com/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow#gaining-consent-for-the-middle-tier-application) to access the server app, so a consent dialog is shown for the server app.
```

Author: good catch!

* Navigate to the server app registration
* In the left hand menu, select **Manifest**
* Replace `"knownClientApplications": []` with `"knownClientApplications": ["<client application id>"]`
* Select **Save**

#### Testing

Ensure you run `azd env set AZURE_USE_AUTHENTICATION` to enable the login UI once you have setup the two Azure AD apps before you deploy or run the application. The login UI will not appear unless all [required environment variables](#environment-variables-reference) have been setup.

> In both the chat and ask a question modes, an optional **Use oid security filter** and **Use groups security filter** will appear. The oid (User ID) filter maps to the `oids` field and the groups (Group ID) filter maps to the `groups` field in the search index. Use the optional scripts included in the sample to manage values for these fields.

Reviewer: ```suggestion
In both the chat and ask a question modes, an optional **Use oid security filter** and **Use groups security filter** will appear. The oid (User ID) filter maps to the `oids` field in the search index and the groups (Group ID) filter maps to the `groups` field. Use the optional scripts included in the sample to manage values for these fields.
```

It's possible that your tenant admin has placed a restriction on consent to apps with [unverified publishers](https://learn.microsoft.com/azure/active-directory/develop/publisher-verification-overview). In this case, only admins may consent to the client and server apps, and normal user accounts are unable to use the login system until the admin consents on behalf of the entire organization.

## Optional scripts

Two optional scripts are provided that allow easier setup of sample data with document level access control.

### Azure Data Lake Storage Gen2 Setup

> [Azure Data Lake Storage Gen2](https://learn.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-introduction) implements an [access control model](https://learn.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-access-control) that can be used for document level access control. The [adlsgen2setup.ps1](./scripts/adlsgen2setup.ps1) script uploads the sample data included in the [data](./data) folder to a Data Lake Storage Gen2 storage account.

Reviewer: ```suggestion
[Azure Data Lake Storage Gen2](https://learn.microsoft.com/azure/storage/blobs/data-lake-storage-introduction) implements an [access control model](https://learn.microsoft.com/azure/storage/blobs/data-lake-storage-access-control) that can be used for document level access control. The [adlsgen2setup.ps1](./scripts/adlsgen2setup.ps1) script uploads the sample data included in the [data](./data) folder to a Data Lake Storage Gen2 storage account.
```

> In order to use this script, an existing Data Lake Storage Gen2 storage account. Run `azd env set AZURE_ADLS_GEN2_STORAGE_ACCOUNT <your-storage-account>` prior to running the script.

Reviewer: First sentence is missing a verb somewhere?

> In order to use this script, an existing Data Lake Storage Gen2 storage account. Run `azd env set AZURE_ADLS_GEN2_STORAGE_ACCOUNT <your-storage-account>` prior to running the script.

Reviewer: Do you need to set up roles/permissions to be able to access the account locally?

Author: Correct - adding this to the doc

To run the script, run the following command: `./scripts/adlsgen2setup.ps1`. The script performs the following steps:
* Creates 3 example [groups](https://learn.microsoft.com/azure/active-directory/fundamentals/how-to-manage-groups): `GPTKB_AdminTest`, `GPTKB_EmployeeTest`, `GPTKB_HRTest`
* Creates a filesystem / container `gptkbcontainer` in the storage account.
* Creates two directories, `benefitinfo` and `employeeinfo` in the `gptkbcontainer` filesystem / container.
* Uploads the sample PDFs into both directories.
* [Recursively sets Access Control Lists (ACLs)](https://learn.microsoft.com/azure/storage/blobs/data-lake-storage-acl-cli) on the `benefitinfo` and `employeeinfo` for the following groups:
  * `GPTKB_AdminTest`: Can read all files in `gptkbcontainer`.
  * `GPTKB_EmployeeTest`: Can only read files in `employeeinfo`.
  * `GPTKB_HRTest`: Can read files in both `employeeinfo` and `benefitinfo`.

In order to use the sample access control, you need to join these groups in your Azure AD tenant.

> Note that this optional scripts may not work in codespaces if your administrator has applied a [Conditional Access policy](https://learn.microsoft.com/azure/active-directory/conditional-access/overview) to your tenant.

Reviewer: ```suggestion
Note that this optional scripts may not work in Codespaces if your administrator has applied a [Conditional Access policy](https://learn.microsoft.com/azure/active-directory/conditional-access/overview) to your tenant.
```

### Azure Data Lake Storage Gen2 Prep Docs

> Once a Data Lake Storage Gen2 storage account has been setup with sample data and access control lists, the [prepdocs.ps1](./scripts/prepdocs.ps1) can be used to automatically process PDFs in the storage account and store them with their [access control lists in the search index](https://learn.microsoft.com/azure/storage/blobs/data-lake-storage-access-control).

Reviewer: ```suggestion
Once a Data Lake Storage Gen2 storage account has been setup with sample data and access control lists, [prepdocs.py](./scripts/prepdocs.py) can be used to automatically process PDFs in the storage account and store them with their [access control lists in the search index](https://learn.microsoft.com/azure/storage/blobs/data-lake-storage-access-control).
```

To run this script with a Data Lake Storage Gen2 account, first set the following environment variables:

* `AZURE_ADLS_GEN2_STORAGE_ACCOUNT`: Name of existing [Data Lake Storage Gen2 storage account](https://learn.microsoft.com/azure/storage/blobs/data-lake-storage-introduction).
* (Optional) `AZURE_ADLS_GEN2_FILESYSTEM`: Name of existing Data Lake Storage Gen2 filesystem / container in the storage account. If empty, `gptkbcontainer` is used.
* (Optional) `AZURE_ADLS_GEN2_FILESYSTEM_PATH`: Specific path in the Data Lake Storage Gen2 filesystem / container to process. Only PDFs contained in this path will be processed.

Once the environment variables are set, run the script using the following command: `./scripts/prepdocs.ps1` or `./scripts/prepdocs.sh`.

### Manually managing Document Level Access Control

> Manually enable document level access control on a search index and manually set access control values using the [manageacl.ps1](./scripts/manageacl.ps1) script.

Reviewer: Is this an alternative to the data lake approach? It might be worth mentioning that at the beginning, just because it's easy to lose track of that in this long readme.

Author: Yes will do that

> Run `azd up` or manually set the `AZURE_SEARCH_SERVICE` and `AZURE_SEARCH_INDEX` environment variables prior to running the script.

Reviewer: Assuming these need to be set as azd env vars -

```suggestion
Run `azd up` or use `azd env set` to manually set `AZURE_SEARCH_SERVICE` and `AZURE_SEARCH_INDEX` prior to running the script.
```

> To run the script, the following parameters are used:

Reviewer: Is this meant to be a parameters reference, or stepbystep of which commands to use? A reference is helpful, but it might also be useful to show an example usage that enables ACLs and adds ACLs for a document, and to clarify that this section is a reference.

Author: This is right below on 161

Author: changed the formatting to make this more clear

* `./scripts/manageacls.ps1 --document [name-of-pdf.pdf] --acl-type [oids or groups]--acl-action remove_all`: Removes all access control values associated with either User IDs or Group IDs for a specific document. Example to remove all Group IDs from the Benefit_Options PDF: `./scripts/manageacls.ps1 --document Benefit_Options.pdf --acl-type groups --acl-action remove_all`.
* `./scripts/manageacls.ps1 --document [name-of-pdf.pdf] --acl-type [oids or groups]--acl-action remove --acl [ID of group or user]`: Removes an access control value associated with either User IDs or Group IDs for a specific document. Example to remove a specific User ID from the Benefit_Options PDF: `./scripts/manageacls.ps1 --document Benefit_Options.pdf --acl-type oids --acl-action remove --acl xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx`.

## Environment Variables Reference

The following environment variables are used to setup the optional login and document level access control:

* `AZURE_USE_AUTHENTICATION`: Enables Azure AD based optional login and document level access control. Set to true before running `azd up`.
* `AZURE_SERVER_APP_ID`: (Required) Application ID of the Azure AD app for the API server.

> * `AZURE_SERVER_APP_SECRET`: [Client secret](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow) used by the API server to authenticate using the Azure AD API server app.

Reviewer: As I mentioned above, we've been encouraged to use certificates instead.

Author: Agreed - will follow up one bicep support is available
